drivers/dma/dma_iproc_pax_v2.c

```diff
@@ -0,0 +1,1111 @@
+		sys_write32(status,
+			    RM_RING_REG(pd, idx,
+					RING_COMPLETION_INTERRUPT_STAT_CLEAR));
+		if (status & 0x1) {
+			k_sem_give(&pd->ring[idx].alert);
+		}
+	}
+}
+#endif
+
+static int dma_iproc_pax_init(const struct device *dev)
+{
+	struct dma_iproc_pax_cfg *cfg = PAX_DMA_DEV_CFG(dev);
+	struct dma_iproc_pax_data *pd = PAX_DMA_DEV_DATA(dev);
+	int r;
+	uintptr_t mem_aligned;
+
+	pd->dma_base = cfg->dma_base;
+	pd->rm_comm_base = cfg->rm_comm_base;
+	pd->used_rings = (cfg->use_rings < PAX_DMA_RINGS_MAX) ?
+			 cfg->use_rings : PAX_DMA_RINGS_MAX;
+
+	/* dma/rm access lock */
+	k_mutex_init(&pd->dma_lock);
+
+	/* Ring Manager H/W init */
+	if (init_rm(pd)) {
+		return -ETIMEDOUT;
+	}
+
+	/* common rm config */
+	rm_cfg_start(pd);
+
+	/* individual ring config */
+	for (r = 0; r < pd->used_rings; r++) {
+		/* per-ring mutex lock */
+		k_mutex_init(&pd->ring[r].lock);
+		/* Init alerts */
+		k_sem_init(&pd->ring[r].alert, 0, 1);
+
+		pd->ring[r].idx = r;
+		pd->ring[r].ring_base = cfg->rm_base +
+					PAX_DMA_RING_ADDR_OFFSET(r);
+		LOG_DBG("RING%d,VERSION:0x%x\n", pd->ring[r].idx,
+			sys_read32(RM_RING_REG(pd, r, RING_VER)));
+
+		/* Allocate for 2 BD buffers + cmpl buffer + payload struct */
+		pd->ring[r].ring_mem = (void *)((uintptr_t)cfg->bd_memory_base +
+					r * PAX_DMA_PER_RING_ALLOC_SIZE);
+		if (!pd->ring[r].ring_mem) {
+			LOG_ERR("RING%d failed to alloc desc memory!\n", r);
+			return -ENOMEM;
+		}
+		/* Find 8K aligned address within allocated region */
+		mem_aligned = ((uintptr_t)pd->ring[r].ring_mem +
+			       PAX_DMA_RING_ALIGN - 1) &
+			       ~(PAX_DMA_RING_ALIGN - 1);
+
+		pd->ring[r].cmpl = (void *)mem_aligned;
+		pd->ring[r].bd = (void *)(mem_aligned +
+					  PAX_DMA_RM_CMPL_RING_SIZE);
+		pd->ring[r].payload = (void *)((uintptr_t)pd->ring[r].bd +
+				      PAX_DMA_RM_DESC_RING_SIZE *
+				      PAX_DMA_NUM_BD_BUFFS);
+
+		LOG_DBG("Ring%d,allocated Mem:0x%p Size %d\n",
+			pd->ring[r].idx,
+			pd->ring[r].ring_mem,
+			PAX_DMA_PER_RING_ALLOC_SIZE);
+		LOG_DBG("Ring%d,BD:0x%p, CMPL:0x%p, PL:0x%p\n",
+			pd->ring[r].idx,
+			pd->ring[r].bd,
+			pd->ring[r].cmpl,
+			pd->ring[r].payload);
+
+		/* Prepare ring desc table */
+		prepare_ring(&(pd->ring[r]));
+
+		/* initialize ring */
+		init_ring(pd, r);
+	}
+
+	/* set ring config done */
+	rm_cfg_finish(pd);
+
+#ifndef CONFIG_DMA_IPROC_PAX_POLL_MODE
+	/* Register and enable RM interrupt */
+	IRQ_CONNECT(DT_INST_IRQN(0),
+		    DT_INST_IRQ(0, priority),
+		    rm_isr,
+		    DEVICE_GET(dma_iproc_pax),
+		    0);
+	irq_enable(DT_INST_IRQN(0));
+#else
+	LOG_INF("%s PAX DMA rings in poll mode!\n", PAX_DMA_DEV_NAME(dev));
+#endif
+	LOG_INF("%s RM setup %d rings\n", PAX_DMA_DEV_NAME(dev),
+		pd->used_rings);
+
+	return 0;
+}
+
+static int dma_iproc_pax_gen_desc(struct dma_iproc_pax_ring_data *ring,
+				  bool is_mega,
+				  uint64_t pci_addr,
+				  uint64_t axi_addr,
+				  uint32_t length,
+				  enum pax_dma_dir dir,
+				  uint32_t **non_hdr_bd_count)
```

Comment:
Why it is pointer to pointer?

Reply:
It should work with single pointer, will change.